### ISSUE TITLE
Feedback homework week 3

### ISSUE THREAD

**Issue by remarcmij** [SarKurd/hyf-javascript3#3]
Hi Sarbast, here is my feedback on your homework for week 3 (folder Saturday HW).

When I load up your `index.html` everything seems to be in working order and your app is responsive too (looks good on an iPhone 6/7/8 screen size). I do see a message `here` in the console, which should not be present in 'production' code.

**1.** You have used the `View` class to render the fixed parts of the web page as well as serving as a base class that provides the `fetchJSON()` and `createAndAppend()` methods. That is generally not a good idea. It violates the principle of [Separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). Now the `Repository` and `Contributor` class, through inheritance, also have access to the `start()` method, which they shouldn't.

It would be better to create yet another class, for instance called `BaseView` that just implements the `fetchJSON()` and `createAndAppend()` shared methods, and have all other classes derive from it.

**2.** Your interpretation of the skeleton class structure is not entirely what I intended. You may have noticed that the class names `Repository` and `Contributor` that I chose are in singular form. I expected that you guys would take that as a hint that each of these classes were supposed to represent **one** repository and **one** contributor respectively. That should also be obvious from the rest of the skeleton. But now I remember the comment you made last time:

> to be honest i didn't read the whole HW, i read the first line only, i didn't want to receive any hint

Frankly, that is not a strategy that I would recommend in general :grin:.

**3.** I see a mix of traditional `for` loops and use of the `.forEach()` method. For consistency, please choose one and stick to it (I prefer `.forEach()` or otherwise `for...of`).

**4.** The sorting of the repository information could be simplified by using the standard string method `.localeCompare()`:

```js
this.repos.sort((a, b) => a.name.localeCompare(b.name));
```

The code that you used obviously works, however it does not take into account local language peculiarities. For instance, suppose you have an array with the following French words (listed in the correct dictionary order):

```js
const frenchWords = [ 'effusion', 'égal', 'emballer' ];
```

The result from sorting:

```js
[ 'effusion', 'emballer', 'égal' ]  // <= your method
[ 'effusion', 'égal', 'emballer' ]  // <= localCompare
```

The `.localeCompare()` method might even be able to sort Arabic words (which I know to have no uppercase letters) in the correct dictionary order (perhaps you can try).

**4.** In your `Contributor` class you are not making full use of the `createAndAppend()` functionality:

```js
const img = this.createAndAppend("img", li);
img.setAttribute("src", userData[i].avatar_url);
```

This could be simplified (as you do in the `View` class) to: 

```js
const img = this.createAndAppend("img", li, { src: userData[i].avatar_url });
```

Bottom line: I think you have grasped the concepts we covered in JS3 well, but please pay attention to detail and consistency. Please read the assignments carefully and apply it meticulously in your implementation.

**Comment by SarKurd** [comment]
Hello Jim,

1- The requirement was creating 3 classes that's why i didn't use this method, i thought you guys want us create no more than 3 :D the code is updated now.

2- To be honest i didn't fully read it but i noticed that through class names, i had in mind to do that but didn't have enough time to change the code i already had, i knew i'd face problems which my take me some time to solve them, so i just used my old code

3- Yep, i'll try to stick with forEach

4- Thanks for that, i didn't know that.

5- Updated the code

Sorry i can't implement number 2 at this moment, if you insist, i'll do it next week.

Thanks for your feedback.

**Comment by remarcmij** [comment]
Hi Sarbast, as I mentioned, not fully reading an assignment is not a good strategy and will get you into problems in the future. Don't bother with modifying your code for issue 2, just keep it in mind for a next time.

**Comment by SarKurd** [comment]
I'll make sure to read carefully next time
Thanks
